### PR TITLE
Fix snapping to segment

### LINKED ISSUE #1929: Cannot snap to edge
QField always snaps to a vertex, regardless the configuration.

Introduced in https://github.com/opengisch/QField/pull/603 (most likely)

### PR DESCRIPTION
**Pull request by m-kuhn**
The code for adding Z/M from snapping results always took the closest vertex to force the snapping result.

This code is now disabled for segments and area, we only take the 2D results in these cases.

Note: we don't do any interpolation of Z/M, this differs from QGIS which assumes a linear interpolation of Z/M along line segments.

Fixes #1929

### CODE REVIEW

**Review by nirvn — APPROVED** [review]

**Comment by m-kuhn** [comment]
@signedav how does this look to you?

@soester / @EulenspiegelTill how does this sound to you?

**Comment by m-kuhn** [comment]
CC @lbartoletti for your particular interest in linear interpolation in combination with snapping too ;-)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for ea14d04cbc0b47873ef51ada149ab3693b0c970d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-caadee35f39f1a4178ba5f3bb613cbddce0a8175-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-caadee35f39f1a4178ba5f3bb613cbddce0a8175-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-caadee35f39f1a4178ba5f3bb613cbddce0a8175-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-caadee35f39f1a4178ba5f3bb613cbddce0a8175-x86.apk)

**Review by signedav — APPROVED** [review]
Looks good

File: src/core/snappingutils.cpp [hunk 1]
```diff
@@ -78,9 +78,11 @@ void SnappingUtils::snap()
   QgsPointLocator::Match match = snapToMap( point );
   mSnappingResult = SnappingResult( match );
 
-  //set point containing ZM if existing
+  //set point containing ZM if we snapped to a point/vertex
   QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( currentLayer() );
-  if ( vlayer && match.layer()
+  if ( match.type() != QgsPointLocator::Type::Edge
+       && match.type() != QgsPointLocator::Type::Area
```
[the remark this answers is not in the document]
Comment (m-kuhn) [reply src/core/snappingutils.cpp:84]: Good point, thanks
Actually we can only enter in here if the vertexIndex is something useful. Which is only the case for `QgsPointLocator::Type::Vertex` and `QgsPointLocator::Type::LineEndPoint` Will adjust :+1:

**Comment by soester** [comment]
@m-kuhn 
Not sure how easy it is to distinguish if the digitizing cross is snapping to a vertex or the segment really close to the vertex. This really depends on the implementation.
Would it be possible to deactivate snapping to segment if not used in a project?
for suverying purposes i think linear interpolation of Z/M along line segments would be the best option!

**Comment by m-kuhn** [comment]
If snapping happens closer to the vertex than [tolerance] it will snap to the vertex. And snapping to edge can be deactivated in the project.

Regarding interpolation I'm half convinced, I'd rather rely on post processing (or have it an option and conciously decide on interpolation strategy, if needed)

**Comment by m-kuhn** [comment]
@qfield-fairy style please

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 475307c30b6a2e74f4029072573727d3911131a9:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-d209282f6973a0e838b8c11f79df77e1c4e3cad9-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-d209282f6973a0e838b8c11f79df77e1c4e3cad9-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-d209282f6973a0e838b8c11f79df77e1c4e3cad9-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-d209282f6973a0e838b8c11f79df77e1c4e3cad9-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for f8020b98dc211e2f371ac054c80a93ab81f4eece:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-37d1a4144ec77eefd13d10462f5cf531aa125fc7-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-37d1a4144ec77eefd13d10462f5cf531aa125fc7-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-37d1a4144ec77eefd13d10462f5cf531aa125fc7-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-37d1a4144ec77eefd13d10462f5cf531aa125fc7-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 79002bb2899777844942dc980ffe82f83a9253a8:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-ad27c08fec1215ffb12a637685d9137694bed897-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-ad27c08fec1215ffb12a637685d9137694bed897-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-ad27c08fec1215ffb12a637685d9137694bed897-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1930-ad27c08fec1215ffb12a637685d9137694bed897-x86.apk)

**Pull request merged**